L1Trigger/TrackFindingTracklet/interface/KFin.h

```diff
@@ -0,0 +1,136 @@
+#ifndef L1Trigger_TrackFindingTracklet_KFin_h
+#define L1Trigger_TrackFindingTracklet_KFin_h
+
+#include "L1Trigger/TrackTrigger/interface/Setup.h"
+#include "L1Trigger/TrackerTFP/interface/DataFormats.h"
+#include "L1Trigger/TrackerTFP/interface/LayerEncoding.h"
+#include "L1Trigger/TrackFindingTracklet/interface/ChannelAssignment.h"
+#include "L1Trigger/TrackFindingTracklet/interface/Settings.h"
+#include "SimTracker/TrackTriggerAssociation/interface/TTTypes.h"
+
+#include <vector>
+
+namespace trklet {
+
+  /*! \class  trklet::KFin
+   *  \brief  Class to emulate transformation of tracklet tracks and stubs into TMTT format
+   *  \author Thomas Schuh
+   *  \date   2021, Dec
+   */
+  class KFin {
+  public:
+    KFin(const edm::ParameterSet& iConfig,
+         const tt::Setup* setup_,
+         const trackerTFP::DataFormats* dataFormats,
+         const trackerTFP::LayerEncoding* layerEncoding,
+         const ChannelAssignment* channelAssignment,
+         const Settings* settings,
+         int region);
+    ~KFin() {}
+
+    // read in and organize input tracks and stubs
+    void consume(const tt::StreamsTrack& streamsTrack, const tt::StreamsStub& streamsStub);
+    // fill output products
+    void produce(tt::StreamsStub& accpetedStubs,
+                 tt::StreamsTrack& acceptedTracks,
+                 tt::StreamsStub& lostStubs,
+                 tt::StreamsTrack& lostTracks);
+
+  private:
+    // truncates double precision of val into base precision, +1.e-12 restores robustness of addition of 2 digitised values
+    double digi(double val, double base) const { return (floor(val / base + 1.e-12) + .5) * base; }
+    // basetransformation of val from baseLow into baseHigh using widthMultiplier bit multiplication
+    double redigi(double val, double baseLow, double baseHigh, int widthMultiplier) const;
+    struct Stub {
+      Stub(const TTStubRef& ttStubRef, int layer, double r, double phi, double z, bool psTilt)
+          : valid_(true), ttStubRef_(ttStubRef), layer_(layer), r_(r), phi_(phi), z_(z), psTilt_(psTilt) {}
+      bool valid_;
```

Comment:
Comments needed to make clear: (1) If r,phi,z are global coords measured w.r.t sector centre etc. or if they are residuals. w.r.t. track; (2) That dPhi & dZ are resolutions (3) that psTilt is true if tilted barrel or endcap PS module.

Reply:
done

Reply:
Are dPhi & dZ the real stub uncertainties, or are they uncertainty * sqrt(12)?

Reply:
those are detector resolution * sqrt(12) plus additional terms.

Reply:
So the dPhi & dZ comments should say "phi uncertainty in rad (det. resolution*sqrt(12) + additional terms) etc.

Reply:
again I won't write that everywhere

Reply:
done